cardano/src/address.rs

```diff
@@ -257,72 +266,186 @@ impl cbor_event::de::Deserialize for Attributes {
+impl TryFromSlice for HashedSpendingData {
     type Error = <Blake2b224 as TryFromSlice>::Error;
     fn try_from_slice(slice: &[u8]) -> ::std::result::Result<Self, Self::Error> {
         Ok(Self::from(Blake2b224::try_from_slice(slice)?))
     }
 }
-impl Deref for Addr {
+impl Deref for HashedSpendingData {
     type Target = <Blake2b224 as Deref>::Target;
     fn deref(&self) -> &Self::Target { self.0.deref() }
 }
-impl AsRef<[u8]> for Addr {
+impl AsRef<[u8]> for HashedSpendingData {
     fn as_ref(&self) -> &[u8] { self.0.as_ref() }
 }
-impl From<Addr> for [u8;Blake2b224::HASH_SIZE] {
-    fn from(hash: Addr) -> Self { hash.0.into() }
+impl From<HashedSpendingData> for [u8;Blake2b224::HASH_SIZE] {
+    fn from(hash: HashedSpendingData) -> Self { hash.0.into() }
 }
-impl From<[u8;Blake2b224::HASH_SIZE]> for Addr {
-    fn from(hash: [u8;Blake2b224::HASH_SIZE]) -> Self { Addr(Blake2b224::from(hash)) }
+impl From<[u8;Blake2b224::HASH_SIZE]> for HashedSpendingData {
+    fn from(hash: [u8;Blake2b224::HASH_SIZE]) -> Self { HashedSpendingData(Blake2b224::from(hash)) }
 }
-impl From<Blake2b224> for Addr {
-    fn from(hash: Blake2b224) -> Self { Addr(hash) }
+impl From<Blake2b224> for HashedSpendingData {
+    fn from(hash: Blake2b224) -> Self { HashedSpendingData(hash) }
 }
-impl FromStr for Addr {
+impl FromStr for HashedSpendingData {
     type Err = <Blake2b224 as FromStr>::Err;
     fn from_str(s: &str) -> ::std::result::Result<Self, Self::Err> {
         Ok(Self::from(Blake2b224::from_str(s)?))
     }
 }
 
+/// A valid cardano Address that is displayed in base58
 #[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone)]
+pub struct Addr(Vec<u8>);
+
+impl Addr {
+    pub fn deconstruct(&self) -> ExtendedAddr {
+        let mut raw = RawCbor::from(&self.0);
+        cbor_event::de::Deserialize::deserialize(&mut raw).unwrap() // unwrap should never fail from addr to extended addr
+    }
+}
+
+impl AsRef<[u8]> for Addr {
+    fn as_ref(&self) -> &[u8] { self.0.as_ref() }
+}
+
+impl TryFromSlice for Addr {
+    type Error = cbor_event::Error;
+    fn try_from_slice(slice: &[u8]) -> ::std::result::Result<Self, Self::Error> {
+        let mut v = Vec::new();
+        // TODO we only want validation of slice here, but we don't have api to do that yet.
+        {
+            let mut raw = RawCbor::from(slice);
+            let _ : ExtendedAddr = cbor_event::de::Deserialize::deserialize(&mut raw)?;
+        }
+        v.extend_from_slice(slice);
+        Ok(Addr(v))
+    }
+}
+
+impl From<ExtendedAddr> for Addr {
+    fn from(ea: ExtendedAddr) -> Self { ea.to_address() }
+}
+
+impl fmt::Display for Addr {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", base58::encode(&self.0))
+    }
+}
+
+impl cbor_event::se::Serialize for Addr {
+    fn serialize<W: ::std::io::Write>(&self, serializer: Serializer<W>) -> cbor_event::Result<Serializer<W>> {
+        serializer.write_bytes(self.as_ref())
+    }
```

Comment:
I don't think this is what you want to write here. Here `write_bytes` will write the CBOR byte object. It will start with the length and then the bytes of the `Vec`. It won't be a valid ExtendedAddr anymore after that.